File: python/solid/interface_segregation.py
import abc
from abc import abstractmethod


class Operation(abc.ABCMeta):
    @abstractmethod
    def print(cls):
        pass

    @abstractmethod
    def scan(cls):
        pass

    @abstractmethod
    def fax(cls):
        pass


class MultiFunctionMachine(Operation):  # all of these methods are okay because this obj can handle them
    def print(cls):
        pass

    def scan(cls):
        pass

    def fax(cls):
        pass


class OldMachine(Operation):  # this will break the ISP rule
    def print(cls):
        pass

    def scan(cls):
        pass  # this is not ok because cant handle this operation

    def fax(cls):
        pass  # this is not ok because cant handle this operation

# should separate the interfaces for solution

File: golang/solid/interface_segregation.go
package solid

type Document struct{}

type Operations interface {
	Print()
	Scan()
	Fax()
}

// this is ok because MultiFunctionMachine can handle all of these operations

type MultiFunctionMachine struct{}

func (mf *MultiFunctionMachine) Print() {
	//ok
}
func (mf *MultiFunctionMachine) Scan() {
	//ok
}
func (mf *MultiFunctionMachine) Fax() {
	//ok
}

type OldMachine struct{}

func (om *OldMachine) Print() {
	//ok
}
func (om *OldMachine) Scan() {
	// OldMachine doesn't support Scan, but it is forced to implement this, and it will break Interface Segregation
}
func (om *OldMachine) Fax() {
	// OldMachine doesn't support Fax, but it is forced to implement this, and it will break Interface Segregation
}

// we should separate the interfaces

type Printer interface {
	Print()
}
type Scanner interface {
	Scan()
}

// and ...

// here each struct can implement interfaces it need and does not force to implement unnecessary
